src/result-async.ts

```diff
@@ -0,0 +1,131 @@
+import { Result, Ok, Err } from './result'
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isPromise<T>(obj: any): obj is Promise<T> {
```

Comment:
Please use `unknown` instead of `any` for function argument. 

I think it would be beneficial to have a few tests for this function. Are there any edge cases that you think that should be tested?

I was initially thinking it would be beneficial to use a 3rd party library, but that would add unjustified bloat to this package. But now that there's a custom-made `isPromise` validator, it would be prudent to ensure it's working as expected (plus to verify regressions in the future).

Reply:
Hey, this method was copied from this repo: [then/is-promise](https://github.com/then/is-promise).

The reason I didn't import it is because of the risks associated with dependencies. Very recently, the `is-promise` repo (which is listed as a dependency to 3.4M projects!) introduced a breaking change which wreaked havoc across a huge number of dependants. 

I think it's trivial enough to just paste here. This implementation coincides with the specification of a `thenable`: an `object` or `function` with a `then` property which is a `function`. It's a looser alternative to "instanceof Promise" which returns false for promises coming from other libs like Bluebird for instance. I don't see any edge-cases and if 3.4M projets use it, I guess there aren't any.

I'm not sure if having tests for this method is necessary to be honest. If you prefer, we can copy those located [here](https://github.com/then/is-promise/blob/master/test.js).

Reply:
As for the `unknown` argument type, I tried but I have a compiler error on the `obj.then` section: `Property 'then' does not exist on type 'object'`.

Reply:
Yup, I agree it makes sense not to use a third-party lib. But for the sake of tracking regressions, there should definitely be some tests for this function.

I tried getting rid of `any`, and got as far as [this](https://www.typescriptlang.org/play/#code/C4TwDgpgBACgNgQwJYDsDyAjAVhAxsKAXigCU8B7AJwBMAeAZ2EtQHMAaKAVxQGsVyA7igB8AKFEAzbviTkUUJPXjIUAKXqYc+ABQA3BHABcXXvyEBKY-rgL6sRKk14CAb1FQolCME6V5oSHIJKGsiQmIAcnJsZwioADJ4kIMoAEJwqBROOBtRAF8ocUlpYFl5RRhKcgBbRQhaABVhbWisY24+QRRLKFbbWCra+nqmqDcPLx8-KG13DzTUvsS5j20KhzUNGJ1W8ygAH32oAIgg3uwwyKkUGTkIveX54-BT4NaAOmAACwh5cKuSmUInNzPlREA) ... but I am still getting type errors and so I think it's justified to keep the `any` in this case.

Reply:
I agree about the tests. I just pushed a commit with extensive tests.